File: frontend/src/services/allergen.ts
import { apiClient } from './api'

export interface Allergen {
  id: number
  household_id: number
  name: string
  created_at: string
  updated_at: string
}

export interface CreateAllergenData {
  name: string
}

export const listHouseholdAllergens = async (householdId: number): Promise<Allergen[]> => {
  const response = await apiClient.get(`/households/${householdId}/allergens`)
  return response.data
}

export const createAllergen = async (householdId: number, data: CreateAllergenData): Promise<Allergen> => {
  const response = await apiClient.post(`/households/${householdId}/allergens`, data)
  return response.data
}

export const deleteAllergen = async (allergenId: number): Promise<void> => {
  await apiClient.delete(`/households/allergens/${allergenId}`)
}

// --- Allergen checking ---
//
// Matching lives on the server (src/services/allergen_service.py), which reuses
// the same normalised ingredient matcher as recipe makeability. Doing it here
// would be a second, weaker implementation that disagrees with the backend —
// the old client-side regex flagged "corn" for "cornstarch", for instance.

export interface InventoryAllergenMatch {
  item_id: number
  name: string
  allergens: string[]
}

export interface AllergenTextMatch {
  text: string
  allergens: string[]
}

/**
 * Allergen matches for every stored item in a household, in one request.
 *
 * Deliberately household-wide rather than per item: a list view can flag all of
 * its cards from a single call instead of one lookup per card. Items with no
 * match are omitted.
 */
export const getInventoryAllergenMatches = async (
  householdId: number
): Promise<InventoryAllergenMatch[]> => {
  const response = await apiClient.get<{ matches: InventoryAllergenMatch[] }>(
    `/households/${householdId}/allergens/inventory-matches`
  )
  return response.data.matches
}

/** Check free text (e.g. an add/edit form's ingredients field) for allergens. */
export const checkTextsForAllergens = async (
  householdId: number,
  texts: string[]
): Promise<AllergenTextMatch[]> => {
  const response = await apiClient.post<{ results: AllergenTextMatch[] }>(
    `/households/${householdId}/allergens/check`,
    { texts }
  )
  return response.data.results
}

File: backend/src/api/v1/allergen.py
"""Allergen API routes."""
from fastapi import APIRouter, HTTPException, status

from src.core.deps import CurrentUserId, DbSession
from src.core.exceptions import AuthorizationError, NotFoundError
from src.schemas.allergen import (
    Allergen,
    AllergenCheckRequest,
    AllergenCheckResponse,
    AllergenCreate,
    InventoryAllergenMatchesResponse,
)
from src.services.allergen_service import AllergenService

router = APIRouter()


@router.post("/{household_id}/allergens", response_model=Allergen, status_code=status.HTTP_201_CREATED)
async def create_allergen(
    household_id: int,
    allergen: AllergenCreate,
    user_id: CurrentUserId,
    db: DbSession,
):
    """Create a new allergen for a household."""
    service = AllergenService(db)
    try:
        return await service.create_allergen(user_id, household_id, allergen)
    except AuthorizationError as e:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(e))


@router.get("/{household_id}/allergens", response_model=list[Allergen])
async def list_allergens(
    household_id: int,
    user_id: CurrentUserId,
    db: DbSession,
):
    """List all allergens for a household."""
    service = AllergenService(db)
    try:
        return await service.list_household_allergens(household_id, user_id)
    except AuthorizationError as e:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(e))


@router.get(
    "/{household_id}/allergens/inventory-matches",
    response_model=InventoryAllergenMatchesResponse,
)
async def list_inventory_allergen_matches(
    household_id: int,
    user_id: CurrentUserId,
    db: DbSession,
):
    """Flag every stored inventory item whose ingredients hit a household allergen.

    One household-wide call rather than one per item, so a list view can flag
    all of its cards without an N+1 of lookups. Items with no match are omitted.
    """
    service = AllergenService(db)
    try:
        matches = await service.match_inventory_items(household_id, user_id)
    except AuthorizationError as e:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(e))
    return InventoryAllergenMatchesResponse(matches=matches)


@router.post("/{household_id}/allergens/check", response_model=AllergenCheckResponse)
async def check_allergens(
    household_id: int,
    body: AllergenCheckRequest,
    user_id: CurrentUserId,
    db: DbSession,
):
    """Check free text (e.g. an add/edit form's ingredients) for household allergens."""
    service = AllergenService(db)
    try:
        results = await service.check_texts(household_id, user_id, body.texts)
    except AuthorizationError as e:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(e))
    return AllergenCheckResponse(results=results)


@router.delete("/allergens/{allergen_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_allergen(
    allergen_id: int,
    user_id: CurrentUserId,
    db: DbSession,
):
    """Delete an allergen."""
    service = AllergenService(db)
    try:
        await service.delete_allergen(allergen_id, user_id)
    except NotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except AuthorizationError as e:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(e))
